luarules/gadgets/unit_collision_damage_behavior.lua

```diff
@@ -0,0 +1,171 @@
+
+--any impulse to damage ratio below this is ignored to save performance
+local minImpulseToDamageRatio = 0.2
+
+local groundCollisionDefID = Game.envDamageTypes.GroundCollision
+local objectCollisionDefID = Game.envDamageTypes.ObjectCollision
+local spGetUnitHealth = Spring.GetUnitHealth
+local spGetUnitVelocity = Spring.GetUnitVelocity
+local spSetUnitVelocity = Spring.SetUnitVelocity
+local mathMin = math.min
+local mathMax = math.max
+
+local fallDamageMultipliers = {}
+local unitsMaxImpulse = {}
+local weaponDefIDImpulses = {}
+local transportedUnits = {}
+local unitMasses = {}
+local weaponDefIgnored = {}
+local unitInertiaCheckFlags = {}
+local gameFrame = 0
+local velocityWatchDuration = 8
+
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	local fallDamageMultiplier = unitDef.customParams.fall_damage_multiplier or 1.0
+	fallDamageMultipliers[unitDefID] = fallDamageMultiplier * fallDamageMagnificationFactor
+	unitsMaxImpulse[unitDefID] = unitDef.mass * maxImpulseMultiplier
+	unitMasses[unitDefID] = unitDef.mass
+end
+
+for name, weaponDefID in pairs(Game.envDamageTypes or {}) do
+	weaponDefIgnored[weaponDefID] = true
+end
+
+for weaponDefID, wDef in ipairs(WeaponDefs) do
+	if wDef.damages and wDef.damages.impulseBoost and wDef.damages.impulseFactor then
+		weaponDefIDImpulses[weaponDefID] = {impulseBoost = wDef.damages.impulseBoost, impulseFactor = wDef.damages.impulseFactor}
+	end
+
+	
+	local function maxDamage(damages)
+		local damage = damages[0]
+		for i = 1, #damages do
+			damage = mathMax(damages[i], damage)
+		end
+		return damage
+	end
+	
+	--generate list of exempted weapons to improve performance
+	if wDef.damages and wDef.damages.impulseFactor == 0 or
+		(wDef.damages.impulseFactor < minImpulseFactor and wDef.damages.impulseBoost < maxDamage(wDef.damages) * minImpulseToDamageRatio) then
+		weaponDefIgnored[weaponDefID] = true
+	end
+end
+
+local function getImpulseMultiplier(unitDefID, weaponDefID, damage)
+	local impulseBoost = weaponDefIDImpulses[weaponDefID].impulseBoost or 0
+	local impulseFactor = weaponDefIDImpulses[weaponDefID].impulseFactor or 1
+	local impulse = (damage + impulseBoost) * impulseFactor
+	local maxImpulse = unitsMaxImpulse[unitDefID]
+	local impulseMultiplier = mathMin(maxImpulse/impulse, 1)
+	return impulseMultiplier
+end
+
+local function massToHealthRatioMultiplier(unitID, unitDefID)
+	local health, maxHealth = spGetUnitHealth(unitID)
+	if maxHealth then
+		local massToMaxHealthMultiplier = (maxHealth / unitMasses[unitDefID]) * fallDamageMultipliers[unitDefID]
+		return massToMaxHealthMultiplier
+	else
+		return fallDamageMultipliers[unitDefID]
+	end
+end
+
+local function isValidCollisionDirection(unitID)
+	local velX, velY, velZ, velLength = spGetUnitVelocity(unitID)
+		-- y-velocity check prevents mostly horizontal object collisions from taking damage, allows damage if dropped from above
+		return velLength > collisionVelocityThreshold and -velY > (velLength * 0.82)
```

Comment:
0.82 looks magic

Reply:
It is magic. It wasn't fun finding this value.

Reply:
how was the value found? it seems to represent a 55° angle of descent, if that was the intention perhaps say so in a comment
![image](https://github.com/user-attachments/assets/acb83bfd-a072-4c17-bc26-3df990d0e7eb)

Reply:
converted to a trigonometric function that converts eular angle into the multiplier for clarity and ease of calibration